Add detail to stream usage examples (#76)
WICG/serial: EXPLAINER.md

In this example we assume a device implementing a protocol inspired by the [Hayes command set](https://en.wikipedia.org/wiki/Hayes_command_set). Since commands are encoded in ASCII a [`TextEncoder`](https://encoding.spec.whatwg.org/#interface-textencoder) and [`TextDecoder`](https://encoding.spec.whatwg.org/#interface-textdecoder) are used to translate the `Uint8Array`s used by the `SerialPort`'s streams to and from strings.

```javascript
const encoder = new TextEncoder();
const writer = encoder.writable.getWriter();
writer.write(encoder.encode("AT"));

const decoder = new TextDecoder();
const reader = port.readable.getReader();
const { value, done } = await reader.read();
console.log(decoder.decode(value));
// Expected output: OK
```

The readable and writable streams must be unlocked before the port can be closed.

```javascript
writer.releaseLock();
reader.releaseLock();
await port.close();
```

Rather than reading a single chunk from the stream code will often read continuously using a loop like this,

```javascript
const reader = port.readable.getReader();

> while (true) {

Reviewer: See my suggestion of using `async for` instead in the other PR.

Author: I've never tried it before. I'd like to stick to simpler examples for the moment.

  const { value, done } = await reader.read();
  if (done) {
    break;
  }
  // Do something with |value|...
}
reader.releaseLock();
```

In this case `port.readable` will not be unlocked until the stream encounters an error, so how do you close the port? Calling `cancel()` on `reader` will cause the `Promise` returned by `read()` to resolve immediately with `{ value: undefined, done: true }`. This will cause the code above to break out of the loop and unlock the stream so that the port can be closed,

```javascript
await reader.cancel();
await port.close();
```

A serial port may generate one of a number of non-fatal read errors for conditions such as buffer overflow, framing or parity errors. These are thrown as exceptions from the `read()` method and cause the `ReadableStream` to become errored. If the error is non-fatal then `port.readable` is immediately replaced by a new `ReadableStream` that picks up right after the error. To expand the example above to handle these errors another loop is added,

```javascript
while (port.readable) {
  const reader = port.readable.getReader();

>   try {

Reviewer: Personally, I'd put the try closer to the `.read()`, but this might be ok just to show how it works and that failing to read() breaks the loop.

Author: I either need to take this suggestion or the one above, because if I use `for await` then I have to wrap the whole loop in a try/catch whereas if I don't then I can put the try only around the call to `read()`.

    }
  } catch (error) {
    // Handle the read |error|...
  }
  reader.releaseLock();
}
```

If a fatal error occurs, such as a USB device being removed, then `port.readable` will be set to `null`.

Revisiting the earlier example, for a device that always produces ASCII text the explicit calls to `encode()` and `decode()` can be removed through the use of [`TransformStream`](https://streams.spec.whatwg.org/#ts)s. In this example `writer` comes from a [`TextEncoderStream`](https://encoding.spec.whatwg.org/#interface-textencoderstream) and `reader` comes from a [`TextDecoderStream`](https://encoding.spec.whatwg.org/#interface-textdecoderstream). The `pipeTo()` method is used to connect these transforms to the port.

```javascript
const encoder = new TextEncoderStream();
const writableStreamClosed = encoder.readable.pipeTo(port.writable);
const writer = encoder.writable.getWriter();
writer.write("AT");

const decoder = new TextDecoderStream();
const readableStreamClosed = port.readable.pipeTo(decoder.writable);
const reader = decoder.readable.getReader();
const { value, done } = await reader.read();
console.log(value);
// Expected output: OK
```

> When piping through a transform stream closing the port becomes more complicated. Closing `reader` or `writer` will cause an error to propagate through the transform streams to the underlying port but it is not immediate. The new `writableStreamClosed` and `readableStreamClosed` promises are required to detect when `port.readable` and `port.writable` have been unlocked. Since canceling `reader` causes the stream to be aborted the resulting error must be caught and ignored,

Reviewer: ```suggestion
When piping through a transform stream closing the port becomes more complicated. Closing `reader` or `writer` will cause an error to propagate through the transform streams to the underlying port. However, this propagation doesn't happen immediately. The new `writableStreamClosed` and `readableStreamClosed` promises are required to detect when `port.readable` and `port.writable` have been unlocked. Since canceling `reader` causes the stream to be aborted the resulting error must be caught and ignored,
```
